### PR TITLE
version

### PR DESCRIPTION
**Pull request by diasdavid**
- [x] cli
- [x] http-api
- [x] core
- [x] tests cli
- [x] tests http-api 
- [x] tests core

### CODE REVIEW

File: src/http-api/index.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+var IPFS = require('../ipfs-core')
+var debug = require('debug')
+var log = debug('api')
+
+exports = module.exports
+
+exports.start = function (callback) {
+  // start IPFS and exports.ipfs = new IPFS()
+
+  exports.ipfs = new IPFS()
```
Comment (Dignifiedquire) [inline src/http-api/index.js]: Hmm, this looks like asking for trouble when the export is only available after `start` was called, also what will happen to module caching plus calling `start` twice?

File: src/http-api/resources/version.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+var ipfs = require('./../index.js').ipfs
+var boom = require('boom')
+
+exports = module.exports
```
Comment (Dignifiedquire) [inline src/http-api/resources/version.js]: Why is this needed?

**Comment by diasdavid** [comment]
going to make it all es6'ish before merging it :)

File: src/http-api/index.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      }
+    }
+  })
+
+  server.connection({
+    port: 9000
+  })
+
+  // load routes
+  require('./routes/version.js')
```
Comment (Dignifiedquire) [inline src/http-api/index.js]: This could be much cleaner:

```js
server.route(require('./routes/version.js')

// version.js
var resources = require('./../resources')

module.exports = {
  method: 'GET',
  path: '/api/v0/version',
  handler: resources.version.get
}
```

File: src/ipfs-core/index.js [hunk 4]
```diff
@@ -41,13 +22,15 @@ function IPFS () {
...
-    repo.config.read(function (err, config) {
-      if (err) {
-        return callback(err)
-      }
-      callback(null, config.Version.Current)
+    repo.exists(function (err, exists) {
+      if (err) { return callback(err) }
+
+      repo.config.get(function (err, config) {
+        if (err) { return callback(err) }
```
Comment (Dignifiedquire) [inline src/ipfs-core/index.js]: Inconsistent style, braces + same line

File: src/http-api/index.js (hunk 1 above)
Comment (diasdavid) [inline src/http-api/index.js]: the .ipfs is to be used by the http-api resources, which are only loaded after start is called (from routes), so there is no conflict. Did I get your question right?

Comment (Dignifiedquire) [inline src/http-api/index.js]: Kind of, but why not just pass these resources through explicitly, this could easily break, or be miss used by someone not understanding these exports

File: src/http-api/resources/version.js (hunk 2 above)
Comment (diasdavid) [inline src/http-api/resources/version.js]: because although they point to the same thing initially, if not binded, once exports changes it will stop pointing to module.exports and therefore none of the functions that gets exported by `exports.` will be exposed

more info and examples:
https://blog.tableflip.io/the-difference-between-module-exports-and-exports/

Comment (Dignifiedquire) [inline src/http-api/resources/version.js]: Hmm but you don't use `module.exports`

Comment (diasdavid) [inline src/http-api/resources/version.js]: not in this case, but it is kind of the same why I always add `{}` to if statements, keeping the code uniform, avoid having to debug someday cause 'I forgot one' or something.

Comment (Dignifiedquire) [inline src/http-api/resources/version.js]: okay

File: src/http-api/index.js (hunk 3 above)
Comment (diasdavid) [inline src/http-api/index.js]: why is that _"much"_ cleaner? If there was more than one route on that file I would have to go and

```JavaScript
server.route(require('./routes/version.js').routeA)
server.route(require('./routes/version.js').routeB)
// etc
```

Meaning that I have to alter the route file and index.js, the way I have, I only have to open and modify one file. Also, having the server object available means I always have server methods at my hand.

File: src/cli/commands/id.js [hunk 5]
```diff
@@ -1,5 +1,8 @@
-var Command = require('ronin').Command
-var IPFS = require('../../ipfs-core')
+const Command = require('ronin').Command
+const IPFS = require('../../ipfs-core')
+const debug = require('debug')
+let log = debug('cli:id')
```
Comment (Dignifiedquire) [inline src/cli/commands/id.js]: Can be `const`

File: src/cli/commands/id.js [hunk 6]
```diff
@@ -11,12 +14,10 @@ module.exports = Command.extend({
...
   },
 
-  run: function (name) {
-    var node = new IPFS()
-    node.id(function (err, id) {
-      if (err) {
-        return console.error(err)
-      }
+  run: name => {
+    let node = new IPFS()
```
Comment (Dignifiedquire) [inline src/cli/commands/id.js]: Why `Let`?

File: src/ipfs-core/index.js (hunk 4 above)
Comment (diasdavid) [inline src/ipfs-core/index.js]: fixed now

**Comment by diasdavid** [comment]
Thank you for the CR @Dignifiedquire :)

File: src/ipfs-core/index.js [hunk 7]
```diff
@@ -1,99 +1,79 @@
...
-    opts.url = process.env.APIURL
-  }
-
-  // var api = ipfsAPIclt(config.url)
-  // extend(self, api)
-
   var repo = new IPFSRepo(config.repoPath)
 
-  self.daemon = function (callback) {
+  this.daemon = (callback) => {
```
Comment (Dignifiedquire) [inline src/ipfs-core/index.js]: You can drop the parens around callback

File: src/ipfs-core/index.js [hunk 8]
```diff
@@ -1,99 +1,79 @@
...
+const IPFSRepo = require('ipfs-repo')
 
 exports = module.exports = IPFS
 exports.config = config
 
 function IPFS () {
-  var self = this
-
-  if (!(self instanceof IPFS)) {
+  if (!(this instanceof IPFS)) {
```
Comment (Dignifiedquire) [inline src/ipfs-core/index.js]: If you do this check anyway, why not go `class IPFS` instead?

File: tests/test-cli/test-version.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+/* globals describe, it */
+
+var expect = require('chai').expect
+var nexpect = require('nexpect')
+
+describe('cli: version', function () {
+  it('get the version', function (done) {
```
Comment (Dignifiedquire) [inline tests/test-cli/test-version.js]: I want an arrow function as well

File: tests/test-cli-offline/test-version.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
+/* globals describe, it */
+
+var expect = require('chai').expect
+var nexpect = require('nexpect')
+
+describe('cli-offline: version', function () {
+  it('get the version', function (done) {
```
Comment (Dignifiedquire) [inline tests/test-cli-offline/test-version.js]: me toooo

File: tests/test-api/test-version.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+/* globals describe, it */
+
+var expect = require('chai').expect
+
+process.env.IPFS_PATH = process.cwd() + '/tests/repo-example'
+var api = require('../../src/http-api')
+
+describe('api: version', function () {
+  it('get the version', function (done) {
```
Comment (Dignifiedquire) [inline tests/test-api/test-version.js]: and what about me?

File: src/http-api/index.js (hunk 3 above)
Comment (Dignifiedquire) [inline src/http-api/index.js]: It's not about that, could also change it to

```js
```

**Comment by Dignifiedquire** [comment]
> Thank you for the CR @Dignifiedquire :)

I see code, I review

File: src/ipfs-core/index.js (hunk 8 above)
Comment (diasdavid) [inline src/ipfs-core/index.js]: I was not sure how classes worked underneath, was afraid that it would cause the same problem of method allocation during object initialisation and take a lot of unnecessary memory, but after checking, it seems there is no reason why to be concerned since it moves all the methods to .prototype of class.

File: src/http-api/index.js (hunk 3 above)
Comment (diasdavid) [inline src/http-api/index.js]: we could DI the server obj (and the IPFS core object), but I kind of like making available by knowing where to pick it from other then move it around. I've been doing this for a while, if it becomes complicated over time, I will reconsider your suggestion :)
